### DISCUSSION TITLE
Disable SSR dynamically / Transfer data from server middleware to Nuxt app

### DISCUSSION THREAD

**Discussion by mkarras** [nuxt/nuxt#27665] [General]
Hey everyone,

I'am looking for a solution to render a page dynamically on server or not based on data from an API.

My application uses "pretty URLs" for SEO and user friendly routes. E.g. instead of the path "`/content/1234-5678-9123-4567`" you can configure `"/my-nice-slug`" or "`/my/custom/path`".

To match all paths the page component `[...slug].vue` inside of the root `/pages` folder exists.

Inside the component the content is rendered based on data of an API. The API returns the info about the page type. E.g. if the page is a content or service page. The template looks simplified like this:
```
<Component :is="component" ... />
```
where the rendered component is based on the page type.

Now there is the problem, that a service page need to be rendered on client only, and the content page on server (when initially called).
But both page types using the same page component.

Now I could implement a server middleware like:
```
try {
    const data = await $fetch<...>('/my-translate-path-api', {
      query: { path: event.path },
    });

    // Disable SSR when page type is service.
    if (data?.pageType === PageType.service) {
      event.context.nuxt = event.context.nuxt || {};
      event.context.nuxt.noSSR = true;
    }
  }
```
to set SSR off if needed, but i also need the data from the API call (`/my-translate-path-api`) in my Nuxt app and i don't want to call the API there again.

Is there a way to transfer the data from server middleware to the Nuxt app?

Or even an alternative way to disable SSR dynamically? E.g. using `defineRouteRules` on page component level like:
```
defineRouteRules({
   ssr: pageData?.type !== PageType.service,
});
```

Using the client-only component to render the client only part is no solution. Because the layout is then rendered on server, but the page content not. This can lead to show the user the layout, executing the page content logic that could result to an error, switching then to the error app.

Thanks for any help in advance.

**Comment by TheAlexLichter** [comment]
FYI: I made a video about [Dynamic Rendering last Friday](https://www.youtube.com/watch?v=vHvfee1PD10)

In your case, you can call and cache the API on the server-side, which minimizes the overhead of another call during SSR ☺️

**Reply by mkarras** [reply]
@manniL thanks for your respond. What do you mean to cache the API?
I found `defineCachedFunction` here https://nitro.unjs.io/guide/cache#cached-functions
Do you mean to create a cached server middleware as a wrapper for the API request (`/my-translate-path-api`) that is then used in another server middleware to disable SSR and again in the Nuxt app?

My current solution is to call the translate-path API inside the server middleware (h3) to disable SSR. But I also need the fetched data inside the Nuxt app. For example the page type to decide wich content component should be rendered.

It would be great to dynamically set SSR off in app code (and call the translate-path API just there).

Or to transfer data from h3 server middleware to the Nuxt app.
